File: parsers/java/src/main/java/com/aperturesyndicate/synx/SynxDiff.java
package com.aperturesyndicate.synx;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Objects;

/** Structural diff between two SYNX objects. Mirrors {@code crates/synx-core/src/diff.rs}. */
public final class SynxDiff {

    private SynxDiff() {}

    public record Change(SynxValue from, SynxValue to) {}

    public record ChangedEntry(String key, Change change) {}

    public static final class Result {
        public final SynxObject added;
        public final SynxObject removed;
        public final List<ChangedEntry> changed;
        public final List<String> unchanged;
        public Result(SynxObject added, SynxObject removed,
                       List<ChangedEntry> changed, List<String> unchanged) {
            this.added = added; this.removed = removed;
            this.changed = changed; this.unchanged = unchanged;
        }
    }

    public static Result diff(SynxObject a, SynxObject b) {
        SynxObject added = new SynxObject();
        SynxObject removed = new SynxObject();
        List<ChangedEntry> changed = new ArrayList<>();
        List<String> unchanged = new ArrayList<>();

        for (var e : a) {
            SynxValue bv = b.get(e.getKey());
            if (bv == null) {
                removed.set(e.getKey(), e.getValue());
            } else if (Objects.equals(e.getValue(), bv)) {
                unchanged.add(e.getKey());
            } else {
                changed.add(new ChangedEntry(e.getKey(), new Change(e.getValue(), bv)));
            }
        }
        for (var e : b) {
            if (a.get(e.getKey()) == null) {
                added.set(e.getKey(), e.getValue());
            }
        }
        Collections.sort(unchanged);
        return new Result(added, removed, changed, unchanged);
    }

    public static SynxValue toValue(Result d) {
        SynxObject root = new SynxObject();
        root.set("added",   SynxValue.ofObject(d.added));
        root.set("removed", SynxValue.ofObject(d.removed));

        SynxObject changedObj = new SynxObject();
        for (ChangedEntry e : d.changed) {
            SynxObject inner = new SynxObject();
            inner.set("from", e.change().from());
            inner.set("to",   e.change().to());
            changedObj.set(e.key(), SynxValue.ofObject(inner));
        }
        root.set("changed", SynxValue.ofObject(changedObj));

        List<SynxValue> arr = new ArrayList<>();
        for (String s : d.unchanged) arr.add(SynxValue.ofString(s));
        root.set("unchanged", SynxValue.ofArray(arr));
        return SynxValue.ofObject(root);
    }
}

File: parsers/swift/Sources/Synx/SynxDiff.swift
// Structural diff between two SYNX values. Mirrors crates/synx-core/src/diff.rs.
import Foundation

public struct SynxDiffChange: Equatable, Sendable {
    public var from: SynxValue
    public var to: SynxValue
}

public struct SynxDiffResult: Sendable {
    public var added: SynxObject
    public var removed: SynxObject
    public var changed: [(key: String, change: SynxDiffChange)]
    public var unchanged: [String]
}

public enum SynxDiff {

    public static func diff(_ a: SynxObject, _ b: SynxObject) -> SynxDiffResult {
        var added = SynxObject()
        var removed = SynxObject()
        var changed: [(String, SynxDiffChange)] = []
        var unchanged: [String] = []

        for entry in a.entries {
            if let bv = b[entry.key] {
                if entry.value == bv {
                    unchanged.append(entry.key)
                } else {
                    changed.append((entry.key, SynxDiffChange(from: entry.value, to: bv)))
                }
            } else {
                removed.set(entry.key, to: entry.value)
            }
        }
        for entry in b.entries where a[entry.key] == nil {
            added.set(entry.key, to: entry.value)
        }
        unchanged.sort()
        return SynxDiffResult(added: added, removed: removed, changed: changed, unchanged: unchanged)
    }

    public static func toValue(_ d: SynxDiffResult) -> SynxValue {
        var root = SynxObject()
        root.set("added", to: .object(d.added))
        root.set("removed", to: .object(d.removed))

        var changedObj = SynxObject()
        for (k, c) in d.changed {
            var inner = SynxObject()
            inner.set("from", to: c.from)
            inner.set("to",   to: c.to)
            changedObj.set(k, to: .object(inner))
        }
        root.set("changed", to: .object(changedObj))

        let arr = d.unchanged.map { SynxValue.string($0) }
        root.set("unchanged", to: .array(arr))
        return .object(root)
    }
}
